lib/excalt/vcard/contact.ex

```diff
@@ -1,83 +1,196 @@
+          addressbook :: String.t(),
+          contanct_url :: String.t()
+        ) :: {:ok, etag :: String.t(), contact :: Contact.t()} | {:error, :not_found} | :error
   def get(server_url, username, password, addressbook, contact_url) do
-    authentication = Excalt.Vcard.Addressbook.build_authentication_header(username, password)
+    authentication = Excalt.Vcard.Helpers.build_authentication_header(username, password)
     req_url = Excalt.Vcard.UrlHelper.build_url(server_url, username, addressbook)
     req_body = Excalt.Vcard.ReqBuilder.get_contact(contact_url)
     req_header = [authentication, {"Depth", "1"}]
     request = Finch.build("REPORT", req_url, req_header, req_body)
 
     case Finch.request(request, ExcaltFinch) do
-      {:ok, %Finch.Response{status: 207, body: body}} ->
-        [contact] = Excalt.Vcard.Parser.parse_contacts_from_addressbook(body)
+      {:ok, %Finch.Response{status: 207, body: body, headers: headers}} ->
+        etag = Excalt.Helpers.Request.extract_from_header(headers, "etag")
+        {:ok, [contact]} = Excalt.Vcard.Parser.parse_contacts_from_addressbook(body)
+        {:ok, etag, contact}
 
       {:ok, %Finch.Response{status: 404, body: _}} ->
         {:error, :not_found}
 
       _ ->
-        # implement more error handling
         :error
     end
   end
 
+  @doc """
+  Creates a new contact, VCF text should be a valid VCARD text
+  If the contact already exists, and we try to do multiple creation of the same contact, the response code 201 is returned but the etag is not changed.
+  """
+  @spec create(
+          server_url :: String.t(),
+          username :: String.t(),
+          password :: String.t(),
+          addressbook_name :: String.t(),
+          vcf_text :: String.t()
+        ) :: {:ok, etag :: String.t()} | {:error, any()}
   def create(server_url, username, password, addressbook_name, vcf_text) do
-    authentication = Excalt.Vcard.Addressbook.build_authentication_header(username, password)
+    authentication = Excalt.Vcard.Helpers.build_authentication_header(username, password)
     req_url = Excalt.Vcard.UrlHelper.build_url(server_url, username, addressbook_name)
     req_body = vcf_text
     contact_uuid = Elixir.UUID.uuid4()
     req_url = "#{req_url}/#{contact_uuid}.vcf"
 
-    request = Finch.build("PUT", req_url, [authentication, {"If-None-Match", "*"}], req_body)
+    request =
+      Finch.build(
+        "PUT",
+        req_url,
+        [authentication, {"If-None-Match", "*"}, {"Content-Type", "text/vcard"}],
+        req_body
+      )
 
     case Finch.request(request, ExcaltFinch) do
-      {:ok, %Finch.Response{status: 201, body: body}} ->
-        {:ok, body}
-
-      _ ->
-        :error
+      {:ok, %Finch.Response{status: 201, body: _body, headers: headers}} ->
+        etag = Excalt.Helpers.Request.extract_from_header(headers, "etag")
+        if etag == "" do
+          {etag, contact} = get_changed_contact_raw(authentication, req_url)
```

Comment:
I would rather return an error instead of fetching the old contact. Otherwise you will loose the creation data.

Reply:
If you find that this is how it should behave, please add a comment with the reference.

Reply:
- This is in fact not way how this library should behave. 
- Responsibility of getting the contact again should be responsibility of the client. 
- Actually it is not an error if the etag is not returned. Returning etag is server responsibility, and on the client is to determine what to do if the etag is missing.